rasa/utils/tensorflow/data_generator.py

```diff
@@ -0,0 +1,428 @@
+from typing import List, Union, Text, Optional, Any, Tuple, Dict
+
+import logging
+import scipy.sparse
+import numpy as np
+import tensorflow as tf
+
+import rasa.shared.utils.io
+from rasa.utils.tensorflow.constants import SEQUENCE, BALANCED
+from rasa.utils.tensorflow.model_data import RasaModelData, Data, FeatureArray
+
+logger = logging.getLogger(__name__)
+
+
+class RasaDataGenerator(tf.keras.utils.Sequence):
+    """Abstract data generator."""
+
+    def __init__(
+        self,
+        model_data: RasaModelData,
+        batch_size: Union[int, List[int]],
+        batch_strategy: Text = SEQUENCE,
+        shuffle: bool = True,
+    ):
+        """Initializes the data generator.
+
+        Args:
+            model_data: The model data to use.
+            batch_size: The batch size(s).
+            batch_strategy: The batch strategy.
+            shuffle: If 'True', data should be shuffled.
+        """
+        self.model_data = model_data
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.batch_strategy = batch_strategy
+
+        self.on_epoch_end()
+
+    def __len__(self) -> int:
+        """Number of batches in the Sequence.
+
+        Returns:
+            The number of batches in the Sequence.
+        """
+        pass
```

Comment:
can this generator be used on itself? should it and other `pass` be `raise NotImplementedError`?